### PR TITLE
make ignore-git true by default when channel is dev

### LINKED ISSUE #43771: Handle git info in rustbuild differently
Right now we encode git information directly into the rustc binary that we build during the normal build process, and this information powers `rustc -vV` and such. At this time, however, this git information is loaded from the git repository every time you execute `./x.py build`. Consequently, whenever a new commit is made, lots of rustc libs need to be rebuilt!

My personal workflow consists of making commits over time as I fix things (especially tests), and having everything get rebuilt every time I make a commit is a bit of a bummer! I think a better strategy here would be to cache git information from the first time of invocation of rustbuild into the out directory. That way we keep reporting the same commit (even though it may not exist or be the actual commit) when there's a persisted build directory.

I believe a strategy like this won't affect the builds on CI, but may hopefully improve the developer experience!

### PR DESCRIPTION
**Pull request by JeremySorensen**
Fixes #43771 
(Handle git info in rustbuild differently)

### CODE REVIEW

**Comment by JeremySorensen** [comment]
This is my first PR, so please tell me what I did wrong!
I wanted to point out that I noticed that the default channel is "dev" when nothing is specified, so I changed git-ignore to true by default to be consistent. (Both in the code and in the config.toml.example file.)

**Comment by Mark-Simulacrum** [comment]
Thanks! If you add "Fixes #43771" to the description (remove issue) GitHub will automatically close it once this merges.

@bors r+

**Comment by ishitatsuyuki** [comment]
I have no idea why you're doing this, as almost any action will cause a full rebuild due to the staged nature.

This removes the version number from my dev channel build, and I will have to change the channel to nightly as the `./configure` script doesn't have option related to git.

Oh, and by the way, dev is the default with git, and stable is the default without git. I think a build should contain version number out of box. If you need to disable for timestamp reasons, just do in your own config.toml.

**Comment by ishitatsuyuki** [comment]
Oh, I see, there was a discussion for this decision. Nevermind my comments.

**Comment by kennytm** [comment]
@bors retry. `distcheck` failed, feels spurious.

```
[00:07:21] downloading https://static.rust-lang.org/dist/2017-07-18/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
######################################################################## 100.0%
[00:07:22] extracting /checkout/obj/build/tmp/distcheck/build/cache/2017-07-18/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:07:22] error: failed to load source for a dependency on `serde_derive`
[00:07:22] 
[00:07:22] Caused by:
[00:07:22]   Unable to update registry https://github.com/rust-lang/crates.io-index
[00:07:22] 
[00:07:22] Caused by:
[00:07:22]   failed to update replaced source `registry https://github.com/rust-lang/crates.io-index`
[00:07:22] 
[00:07:22] Caused by:
[00:07:22]   failed to read root of directory source: /checkout/obj/build/tmp/distcheck/src/vendor
[00:07:22] 
[00:07:22] Caused by:
[00:07:22]   No such file or directory (os error 2)
[00:07:22] failed to run: /checkout/obj/build/tmp/distcheck/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/obj/build/tmp/distcheck/src/bootstrap/Cargo.toml --locked --frozen
[00:07:22] Build completed unsuccessfully in 0:00:09
[00:07:22] Makefile:54: recipe for target 'check' failed
```

**Comment by TimNN** [comment]
`distcheck` failed again, I don't think this is spurious. Looking at the patch, I'm not sure if it entirely correct, although I'm not currently that familiar with the inner working of rustbuild:

This patch changes the default of `ignore-git` to `true` and then later updates the default based on `channel`, but **only if `config.toml` contains a `[rust]` section**. Given that the builders still use `config.mk` I believe that `ignore-git` will now be true on all builders, which then causes the above failure.

File: config.toml.example [hunk 1]
```diff
@@ -259,7 +259,10 @@
 #codegen-tests = true
 
 # Flag indicating whether git info will be retrieved from .git automatically.
-#ignore-git = false
+# Having the git information can cause a lot of rebuilds during development.
+# Note: If this attribute is not explicity set (e.g. it left commented out) it
```
Comment (TimNN) [inline config.toml.example]: This should probably be "**if** left commented out".

Comment (JeremySorensen) [inline config.toml.example]: yes you are right

**Comment by JeremySorensen** [comment]
> Given that the builders still use config.mk I believe that ignore-git will now be true on all builders, which then causes the above failure.

Oh darn... OK well I think the fix is simple, just revert the initial value of ignore-git back to false, and only keep the logic for setting the default if there is a config.toml file.
But that raises a question, are there some tests that:
1. Use a config.toml file
2. Do NOT specify an explicit value for ignore-git
3. Require ignore-git to be false in order to pass?

If all of those are true, for any test, then that(those) test(s) will fail. So if we want to keep this patch, we would need to find all such tests and explicitly set ignore-git to false.

**Comment by JeremySorensen** [comment]
sorry I was trying to figure out how to update a pull request, I didn't realize that pushing to my fork was enough to cause stuff to run. I will follow contributing.md more closely and see if I can figure out what is going on with the tests.

**Comment by JeremySorensen** [comment]
OK I am getting to the point where I need some mentoring here. When I ran the tests (on my machine), a whole block of them failed with the same error. Here is an excerpt of the results:
```
    Check compiletest suite=run-make mode=run-make (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)
    Building rustdoc for stage2 (x86_64-pc-windows-msvc)

    running 159 tests
    test [run-make] run-make\allow-non-lint-warnings-cmdline ... FAILED
    test [run-make] run-make\alloc-extern-crates ... FAILED
    ...
    test [run-make] run-make\windows-subsystem ... FAILED

    failures:

    ---- [run-make] run-make\allow-non-lint-warnings-cmdline stdout ----
    	thread '[run-make] run-make\allow-non-lint-warnings-cmdline' panicked at 'failed to spawn `make`:     Error { repr: Os { code: 2, message: "The system cannot find the file specified." } }', src\libcore\result.rs:860:4
    note: Run with `RUST_BACKTRACE=1` for a backtrace.

    ---- [run-make] run-make\alloc-extern-crates stdout ----
    	thread '[run-make] run-make\alloc-extern-crates' panicked at 'failed to spawn `make`: Error { repr: Os { code: 2, message: "The system cannot find the file specified." } }', src\libcore\result.rs:860:4
    ...
    ---- [run-make] run-make\windows-subsystem stdout ----
	thread '[run-make] run-make\windows-subsystem' panicked at 'failed to spawn `make`: Error { repr: Os { code: 2, message: "The system cannot find the file specified." } }', src\libcore\result.rs:860:4

    failures:
        [run-make] run-make\a-b-a-linker-guard
        [run-make] run-make\alloc-extern-crates
    ...
        [run-make] run-make\windows-subsystem

    test result: FAILED. 0 passed; 159 failed; 0 ignored; 0 measured; 0 filtered out
```
But the thing is I get this even after I revert back to the original config.rs. So the only thing that is changed is config.toml, but everything builds fine its the tests that fail. So I am not sure why these tests are failing but it doesn't seem related to my code.

Also I am not sure why the travis-ci failed, it kind of looks like it got cancelled, if somebody cancelled it because of my comment indicating that I had kicked it off inadvertently then that is fine, but I don't know how to tell that.

At this point the only change to the code is one line:
`config.ignore_git = config.channel == "dev";`
which goes right before this existing line:
`set(&mut config.ignore_git, rust.ignore_git);`
So I think that is right, if the travis-ci was cancelled, can restart it and rerun the tests on the server to see which fail?
Also how can I get those tests to pass? Should I just do a clean clone of my fork and start over?

**Comment by Mark-Simulacrum** [comment]
It looks like you might not have `make` installed. Maybe try running that and see if it exists? Travis passed, and I think the code is correct, so:

@bors r+

**Comment by alexcrichton** [comment]
@bors: r- 

Oh actually mind resting one more thing? I'm worried that with this change when the channel is set via the configure script it will ignore git. Can you run from a clean dir with `./configure --release-channel=nightly`? Or at least I think that is the name

**Comment by JeremySorensen** [comment]
@alexcrichton `CONTRIBUTING.md` says this:

> Note: Previously the `./configure` script was used to configure this project. It can still be used, but it's recommended to use a `config.toml` file. If you still have a `config.mk` file in your directory - from `./configure` - you may need to delete it for config.toml to work.

So it seems like if you are using `./configure` you aren't using `config.toml`. The code to set the default is inside a block like this:
```rust
if let Some(ref rust) = toml.rust {
    ...
    config.ignore_git = config.channel == "dev";
    set(&mut config.ignore_git, rust.ignore_git);
    ...
}
```
So if they don't have the `config.toml` file, the code won't be run. (I am assuming that without `config.toml`, `toml.rust` is guaranteed to be `None`, @Mark-Simulacrum can you confirm that?)
Now if the user has the `config.toml` file, but is trying to use `./configure` then you are right, things would break, but again `CONTRIBUTING.md` sort of says that isn't a correct way to do things.

In any case I would like some clarification because I don't know if you want me to run the `./configure` script with or without the `config.toml` file (Although I guess "clean dir" implies "without", in which case I don't think it is necessary because of what I stated above).

I also have a sort of procedural question, when you asked me to run in a clean dir, do you mean on my own machine? If so, it seems like if we need to check that running with the release channel works correctly, we should add a real test. Otherwise the check I do at home will be lost, both the result and the fact that I ever did it (except in these comments I guess). I am having trouble with some tests due to `make` on Windows, so I will try running all this in an Ubuntu VirtualBox instance and see if the tests pass.

Sorry for all the comments and questions, the ramifications of this issue go further than I thought they would when I started.

**Comment by alexcrichton** [comment]
Oh heh, I also missed the `if let Some(ref rust)`!

But yeah sorry, lemme clarify. Yes most contributors aren't using `./configure`, but all our CI builders do (it's how we produce releases), and it's crucially important that all our releases continue to use git information. Now that being said, testing locally, looks like `./configure --release-channel=beta` still works but that's because of the bug of only working with `[rust]`. A fresh `./x.py build` also has this disabled when we want it enabled by default there.

> In any case I would like some clarification because I don't know if you want me to run the ./configure script with or without the config.toml file (Although I guess "clean dir" implies "without", in which case I don't think it is necessary because of what I stated above).

Oh no worries, I hope the above words clarified!

**Comment by JeremySorensen** [comment]
@alexcrichton Are you saying that you have already `run ./x.py build` on this code and found that the git information was missing? That would be confusing to me (although this is all a little confusing). One question I would ask is, did you run it since I made the second commit? I ask because before I knew about things being run without the `config.toml` file, I changed the initial value of `git-ignore` from `false` to `true` which we found is definitely wrong when using `./configure` but I have since changed it back so we shouldn't have that problem now.
If it sounds like I am subtly trying to get you to run the test for me, I'm not really, but I am concerned that once I get this up and running in Linux, I won't be able to tell if we are running into this issue or not. I hope that if I run the tests and they pass that is enough to show that the information is there. Also, are there any instructions for using `./configure` instead of `config.toml`? Is this sufficient?
```
./configure --release-channel=release
./x.py build
./x.py test > test.results
```

**Comment by JeremySorensen** [comment]
@Mark-Simulacrum By the way I installed `make` and it made some tests pass, but most failed with weird errors like 
> process_begin: CreateProcess(NULL, uname, ...) failed.
The syntax of the command is incorrect. 

So I am going to try doing everything on Linux in a VirtualBox VM.

**Comment by alexcrichton** [comment]
@JeremySorensen ah yeah I checked out this branch and tested it out locally. As-is if you start from a fresh checkout and run `./x.py build` then it includes git information, but it shouldn't ideally! That's at least the first bug to fix I believe.

**Comment by JeremySorensen** [comment]
@alexcrichton Now I am even more confused. I thought if you start from a fresh checkout and run `./x.py build` it should have the git information otherwise the tests that use make will fail. That was why I changed the code so that it only changes the default if you have a `config.toml` present.

**Comment by alexcrichton** [comment]
Ah so "morally" what should happen here is that the value of `ignore-git`, if not otherwise specified, should be inferred from the value of `channel`. The value of `channel` isn't known, however, until we've read in both `config.toml` and `config.mk`. The latter is generated by `./configure` right now where the former is only generated by humans currently.

Currently the `ignore-git` is inferred *only* if `[rust]` present, but the channel is "dev" in many more situations than that, for example with no `config.mk` or no `config.toml`.

**Comment by JeremySorensen** [comment]
@alexcrichton Sorry if I am being dense. I think I might finally understand the situation here.
I am going to try to explain my thought process chronologically and see if I end up in the same place as you:
1. First I thought users could accept the default configuration or change it with `config.toml`. (I didn't know about the other ways to configure it.) Since the default channel is `"dev"` I initialized the default value of `ignore-git` to `true`. (And set the default value in the [rust] block code).
2. This caused an error because some tests didn't use `config.toml` but needed `ignore-git` to be `false`. I incorrectly assumed that these tests just used the default configuration. So I concluded that these tests had the channel set to `"dev"` (because that's the default) but needed `ignore-git` to be `false` anyway.
3. It sounds like in fact these tests specify the channel to be other than `"dev"` by a different means than `config.toml` and so should have `ignore-git` set to `false`. So I should initialize `ignore-git` to `true` again for the case that no configuration is specified, but should also account for ANY form of configuration that changes the channel when adjusting the default value of `ignore-git`.

So my first question is, do any tests or releases expect the channel to be `"dev"` but also expect `ignore-git` to be false? This seems possible because that is the what would happen by default before this PR.

My next question is, assuming I am finally starting to understand what we are going for, would it make sense to introduce a flag that indicates if the user explicitly set `ignore-git` to a specific value? Then at the very end we could change the value of `ignore-git` based on the channel only if that flag is `false`.
To do that all I need to do is go through `config.rs` and find all the places where `ignore-git` is set (not counting initialization) and set that flag to `true` at each point. This seems like a simple text search.
Would that work?
Another option instead of a flag would be to change `ignore-git` to an enum with variants like `Ignore`, `DoNotIgnore` and `NotSet`. I would initialize it to `NotSet` and then if it is still `NotSet` by the end, change it to `Ignore` or `DoNotIgnore` based on the channel. (Is there a standard `TriState` enum with `True`, `False` and `NotSet` already?)

Again sorry for being so long winded (long fingered?) and thanks for your help and patience.

**Comment by alexcrichton** [comment]
@JeremySorensen oh no worries! I think you're spot on in your analysis. That's sort of along the lines of [what I originally tried](https://github.com/rust-lang/rust/pull/43873) where I used `Option<bool>` as a "tri-state enum". What we'll want though is that `Config` itself should continue to have a `bool` whereas the local parsing may wish to have a tri-state boolean. 

Maybe something like:

* When parsing, we store `Option<bool>`
* Configurations of `ignore_git` will set this `Option<bool>`
* At the end we have: `self.ignore_git = local_bool.unwrap_or(self.channel == "dev")`

**Comment by alexcrichton** [comment]
@JeremySorensen I've landed a few tweaks to `config.toml` and such which actually now basically just deletes `config.mk`. I think it may be easier to apply this change now, and you can see how things like `debug_assertions` are handled and probably just mirror that.

**Comment by alexcrichton** [comment]
@bors: r+

Looks great to me, thanks @JeremySorensen!

**Comment by JeremySorensen** [comment]
@alexcrichton does using the "r+" with bors mean that the PR will be merged into `master` if the tests pass? If so, we might want to hold off; frankly I am not sure anymore what the text of `config.toml.example` should be, but I think it is probably out of date with respect to these latest changes. If you are OK with it I am too, but it's easy to forget about the comments in a file that isn't actually run, so I thought I would bring it up.
The issue I have with the language is that the commented out part is supposed to be the default value, but the default value depends on the channel, so it is complicated.

**Comment by kennytm** [comment]
Unable to update registry, spurious. But @JeremySorensen wants to hold off a bit?

**Comment by alexcrichton** [comment]
@JeremySorensen oh bors is our integration bot, when we "r+" a PR it goes [into a queue](https://buildbot2.rust-lang.org/homu/queue/rust) of PRs to be tested, and then the queue is drained one at a time. You've also now seen what happens when tests fail! (spuriously)

The comment looks fine by me though, but @JeremySorensen did you want to reword it?

**Comment by JeremySorensen** [comment]
@alexcrichton I think the expected behavior is that if the user doesn't copy `config.toml.example` over, OR if they do but don't change it, then the channel will be `"dev"` and `ignore_git` will be `true` correct? If so, then yes the comment looks correct to me. (I thought I remembered it was still set to `false` in the example, but I was wrong.)

**Comment by alexcrichton** [comment]
Ok! 

I think, though, that the failure here is not a spurious failure. I believe what you'll want to do is add an option to the configure script for this one builder. You can find the builder's configuration in `src/ci/docker/x86_64-gnu-distcheck` and the env var should be something like `RUST_CONFIGURE_ARGS`. The argument you want to add is `--set build.ignore-git=true`

Note that `--set` hasn't been tested much before, so there may be a bug or two with that...

**Comment by alexcrichton** [comment]
Er sorry, you want `ignore-git=false` b/c we want that builder to read git information!

**Comment by arielb1** [comment]
```
configure: processing command line

[00:00:27] configure: 

[00:00:27] configure: rust.debug-assertions := True

[00:00:27] configure: build.ignore-git     := false

[00:00:27] configure: build.submodules     := False

[00:00:27] configure: llvm.assertions      := True

[00:00:27] configure: build.locked-deps    := True

[00:00:27] configure: llvm.ccache          := sccache

[00:00:27] configure: build.openssl-static := True

[00:00:27] configure: build.build          := x86_64-unknown-linux-gnu

[00:00:27] configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--set',  ...

[00:00:27] Traceback (most recent call last):

[00:00:27]   File "/checkout/src/bootstrap/configure.py", line 378, in <module>

[00:00:27]     configure_section(sections[section_key], section_config)

[00:00:27]   File "/checkout/src/bootstrap/configure.py", line 367, in configure_section

[00:00:27]     raise RuntimeError("failed to find config line for {}".format(key))

[00:00:27] RuntimeError: failed to find config line for ignore-git
```

@bors r-

**Comment by alexcrichton** [comment]
I think maybe `rust.ignore-git`? @JeremySorensen I'm shooting the dark here so it'd be best if you could verify locally, which I believe you can do via `./src/ci/docker/run.sh x86_64-gnu-distcheck`

**Comment by JeremySorensen** [comment]
I ran it as is and got the same error (as expected), then I changed the line in `Dockerfile` to 
`ENV RUST_CONFIGURE_ARGS --build=x86_64-unknown-linux-gnu --set rust.ignore-git=false`
now I get this:
> failed to parse TOML configuration 'config.toml': invalid type: string "false", expected a boolean for key `rust.ignore-git`

I checked, the false is not in quotes in the Docker file, so I guess it is getting escaped incorrectly, or not getting parsed.

**Comment by alexcrichton** [comment]
I think we may need to add a special case to [this line](https://github.com/rust-lang/rust/blob/630e02f25be1e65b316857c5bd8022da0b96db40/src/bootstrap/configure.py#L254) to just convert `"true"` to `True` to get properly set as a boolean perhaps?

**Comment by JeremySorensen** [comment]
@alexcrichton Ha! I just did that! my code looks like this:
```python
for key in known_args:
    # The `set` option is special and can be passed a bunch of times
    if key == 'set':
        for option, value in known_args[key]:
            keyval = value.split('=', 1)
            if len(keyval) == 1 or keyval[1] == "true":
                value = True
            elif keyval[1] == "false":
                value = False
            else:
                value = keyval[1]
            set(keyval[0], value)
        continue
```
We actually need the false case here
Should I commit that for this PR as well?

**Comment by alexcrichton** [comment]
Looks good to me!

**Comment by JeremySorensen** [comment]
I ran the tests again and it looks like it got the ignore-git=false, but I am having a bunch of seemingly unrelated errors. I don't know anything about Docker (I had to install it to make the command work) but I wonder if it is because I ran the test several times and killed it in progress a couple times, or if somehow I really broke everything. Should I push the changes to the Docker file and configure.py?

I didn't redirect the output of the command to a file and it rolled off the end so I don't know if I got everything but here are at least some of the errors:
```
get_toml (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_toml ... ok
program_config (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.program_config ... FAIL
rustc_stamp (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.rustc_stamp ... ok
...
Return True when the stamp file does not exists ... ok

======================================================================
FAIL: program_config (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.program_config
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/doctest.py", line 2226, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for bootstrap.RustBuild.program_config
  File "/checkout/src/bootstrap/bootstrap.py", line 519, in program_config

----------------------------------------------------------------------
File "/checkout/src/bootstrap/bootstrap.py", line 527, in bootstrap.RustBuild.program_config
Failed example:
    cargo_path.rstrip(".exe") == os.path.join("/tmp/rust",
    "bin", "cargo")
Expected:
    True
Got:
    False


----------------------------------------------------------------------
Ran 14 tests in 0.068s

FAILED (failures=1)
invalid checksum:
    found:    334d016f755cd6dc58c53a86e183882f8ec14f52fb05345887c8a5edd42c87b7
    expected: 64ec88ca00b268e5ba1a35678a1b5316d212f4f366b2477232534a8aeca37f3c
+ python2.7 ../x.py test distcheck
    Finished dev [unoptimized] target(s) in 0.0 secs
Distcheck
...
failures:
    net::tcp::tests::clone_accept_concurrent
    net::tcp::tests::clone_accept_smoke
    net::tcp::tests::clone_while_reading
    net::tcp::tests::close_read_wakes_up
    net::tcp::tests::close_readwrite_smoke
    net::tcp::tests::connect_loopback
    net::tcp::tests::double_bind
    net::tcp::tests::fast_rebind
    net::tcp::tests::multiple_connect_interleaved_greedy_schedule
    net::tcp::tests::multiple_connect_interleaved_lazy_schedule
    net::tcp::tests::multiple_connect_serial
    net::tcp::tests::partial_read
    net::tcp::tests::peek
    net::tcp::tests::read_eof
    net::tcp::tests::shutdown_smoke
    net::tcp::tests::smoke_test
    net::tcp::tests::socket_and_peer_name
    net::tcp::tests::tcp_clone_smoke
    net::tcp::tests::tcp_clone_two_read
    net::tcp::tests::tcp_clone_two_write
    net::tcp::tests::write_close
    net::udp::tests::connect_send_peek_recv
    net::udp::tests::peek_from
    net::udp::tests::set_nonblocking
    net::udp::tests::socket_name_ip4
    net::udp::tests::socket_smoke_test_ip4
    net::udp::tests::udp_clone_smoke
    net::udp::tests::udp_clone_two_read
    net::udp::tests::udp_clone_two_write

test result: FAILED. 773 passed; 29 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by alexcrichton** [comment]
Yeah that all looks fine to me, thanks for investigating though! I'm not sure what's up with the python looking failure, but the docker failures I believe are all related to IPv6 in the container. I get them locally as well and I'm not sure how to fix them :(

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**